File: BaekjoonAlgorithm/202102/BA_15686/BJAlgo_15686.py
from sys import stdin
from itertools import combinations


def find_min_dist(chicken):
    min_dist = 0
    for hx, hy in home:
        min_dist_per_home = 101
        for cx, cy in chicken:
            dist = abs(hx - cx) + abs(hy - cy)
            min_dist_per_home = min(min_dist_per_home, dist)
        min_dist += min_dist_per_home

    return min_dist


n, m = map(int, stdin.readline().split())
city = []
home = []
chickens = []
for i in range(n):
    tmp = list(map(int, stdin.readline().split()))
    for j in range(n):
        if tmp[j] == 1:
            home.append([i, j])
        elif tmp[j] == 2:
            chickens.append([i, j])

    city.append(tmp)

result = -1

for combi in combinations(chickens, m):
    result = find_min_dist(combi) if result == -1 else min(result, find_min_dist(combi))

print(result)

File: BaekjoonAlgorithm/202102/BA_15686/BJAlgo_15686.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.StringTokenizer;

public class BJAlgo_15686 {
    private static ArrayList<int[]> home = new ArrayList<>();
    private static ArrayList<int[]> chickens = new ArrayList<>();

    private static ArrayList<ArrayList<int[]>> combi = new ArrayList<>();

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        for (int i = 0; i < n; i++) {
            int[] city = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            for (int j = 0; j < n; j++) {
                if (city[j] == 1) {
                    home.add(new int[]{i, j});
                } else if (city[j] == 2) {
                    chickens.add(new int[]{i, j});
                }
            }
        }

        combinations(new boolean[chickens.size()], 0, m);

        int result = -1;
        for (ArrayList<int[]> c : combi) {
            if (result == -1) {
                result = findMinDist(c);
            } else {
                result = Math.min(result, findMinDist(c));
            }
        }

        System.out.println(result);
    }

    private static void combinations(boolean[] visited, int start, int r) {
        if (r == 0) {
            ArrayList<int[]> tmp = new ArrayList<>();
            for (int i = 0; i < visited.length; i++) {
                if (visited[i]) {
                    tmp.add(chickens.get(i));
                }
            }
            combi.add(tmp);
            return;
        } else {
            for (int i = start; i < visited.length; i++) {
                visited[i] = true;
                combinations(visited, i + 1, r - 1);
                visited[i] = false;
            }
        }

    }

    private static int findMinDist(ArrayList<int[]> chicken) {
        int minDist = 0;
        for (int[] h : home) {
            int minDistPerHome = 101;
            for (int[] c : chicken) {
                int dist = Math.abs(h[0] - c[0]) + Math.abs(h[1] - c[1]);
                minDistPerHome = Math.min(minDistPerHome, dist);
            }
            minDist += minDistPerHome;
        }

        return minDist;
    }
}
